src/Binding.lua

```diff
@@ -4,146 +4,156 @@ local Type = require(script.Parent.Type)
+		assert(typeof(predicate) == "function", "Expected arg #1 to be a function")
+	end
 
-	newValue = internalData.valueTransform(newValue)
+	local impl = {}
 
-	internalData.value = newValue
-	internalData.changeSignal:fire(newValue)
-end
-
---[[
-	Subscribe to a binding's change signal. This is only accessible by Roact.
-]]
-function Binding.subscribe(binding, handler)
-	local internalData = binding[InternalData]
-
-	--[[
-		If this binding is mapped to another and does not have any subscribers,
-		we need to create a subscription to our source binding so that updates
-		get passed along to us
-	]]
-	if internalData.upstreamBinding ~= nil and internalData.subscriberCount == 0 then
-		internalData.upstreamDisconnect = Binding.subscribe(internalData.upstreamBinding, function(value)
-			Binding.update(binding, value)
+	function impl.subscribe(callback)
+		return BindingInternalApi.subscribe(upstreamBinding, function(newValue)
+			callback(predicate(newValue))
 		end)
 	end
 
-	local disconnect = internalData.changeSignal:subscribe(handler)
-	internalData.subscriberCount = internalData.subscriberCount + 1
+	function impl.update(newValue)
+		-- This operation doesn't make sense; we can leave it as a no-op
+	end
 
-	local disconnected = false
+	function impl.getValue()
+		return predicate(upstreamBinding:getValue())
+	end
 
-	--[[
-		We wrap the disconnect function so that we can manage our subscriptions
-		when the disconnect is triggered
-	]]
-	return function()
-		if disconnected then
-			return
+	return setmetatable({
+		[Type] = Type.Binding,
+		[BindingImpl] = impl,
+	}, BindingPublicMeta)
+end
+
+function BindingInternalApi.join(upstreamBindings)
+	if config.typeChecks then
+		assert(typeof(upstreamBindings) == "table", "Expected arg #1 to be of type table")
+
+		for key, value in pairs(upstreamBindings) do
+			if Type.of(value) ~= Type.Binding then
+				local message = (
+					"Expected arg #1 to contain only bindings, but key %s had a non-binding value"
+				):format(
+					tostring(key)
+				)
+				error(message, 2)
+			end
 		end
+	end
 
-		disconnected = true
-		disconnect()
-		internalData.subscriberCount = internalData.subscriberCount - 1
-
-		--[[
-			If our subscribers count drops to 0, we can safely unsubscribe from
-			our source binding
-		]]
-		if internalData.subscriberCount == 0 and internalData.upstreamDisconnect ~= nil then
-			internalData.upstreamDisconnect()
-			internalData.upstreamDisconnect = nil
+	local impl = {
+		disconnectMethods = nil,
+	}
+
+	local function getValue()
+		local value = {}
+
+		for key, upstream in pairs(upstreamBindings) do
+			value[key] = upstream:getValue()
 		end
+
+		return value
 	end
-end
 
---[[
-	Create a new binding object with the given starting value. This
-	function will be exposed to users of Roact.
-]]
-function Binding.create(initialValue)
-	local binding = {
-		[Type] = Type.Binding,
+	function impl.subscribe(callback)
```

Comment:
Would it make more sense to only subscribe to `upstreamBindings` once and create/destroy the upstream connections depending on if there are subscribers or not? The prior implementation does this.

Upstream connections could easily add up, considering bindings mapped from this one will invoke `subscribe` when subscribed to. A good example of this potentially becoming a problem would be when animating transparency.

Reply:
Just got back into working on this!

I don't think it's a big deal to pass subscriptions up through to their parents. At the end of the day, someone is going to have to track the connection, and consolidating that at the top is the simplest approach!